routes/auth/auth.routes.js

```diff
@@ -0,0 +1,141 @@
+      passwordField: 'password',
+      passReqToCallback: true, // allows us to access req in the call back
+    },
+    async (req, username, password, done) => {
+      username = req.sanitize(username);
+      password = req.sanitize(password);
+      // Check if user and password is valid
+      let user = await User.findOne({ username });
+      let passwordValid = user && bcryptjs.compareSync(password, user.passwordHash);
+      // If password valid call done and serialize user.id to req.user property
+      if (passwordValid) {
+        return done(null, {
+          id: user.id,
+          name: user.username,
+        });
+      }
+      return done(null, false);
+    }
+  )
+);
+
+passport.serializeUser(function (user, cb) {
+  process.nextTick(function () {
+    cb(null, { id: user.id, username: user.username });
+  });
+});
+
+passport.deserializeUser(function (user, cb) {
+  process.nextTick(function () {
+    return cb(null, user);
+  });
+});
+
+// Login Routes
+router.get('/login', isLoggedOut, function (req, res, next) {
+  try {
+    res.render('auth/login', { message: req.flash('error') });
+  } catch (err) {
+    console.error('Sorry, there was an error: ', err);
+    res.render('error');
+  }
+});
+
+router.post(
+  '/login',
+  passport.authenticate('local-login', {
+    successReturnToOrRedirect: '/dashboard',
+    failureRedirect: '/login',
+    successFlash: true,
+    failureFlash: true,
+    successFlash: 'Succesfull!',
+    failureFlash: 'Invalid username or password.',
+  })
+);
+
+// Signup Routes
+router.get('/signup', isLoggedOut, (req, res, next) => {
+  try {
+    res.render('auth/signup');
+  } catch (err) {
+    console.error('Sorry, there was an error: ', err);
+    res.render('error');
+  }
+});
+
+router.post('/signup', async (req, res, next) => {
+  const username = req.sanitize(req.body.username);
+  const password = req.sanitize(req.body.password);
+  if (!username || !password) {
+    res.render('auth/signup', { errorMessage: 'All fields are mandatory. Please provide your username, email and password.' });
+    return;
+  }
+
+  const regex = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}/;
+  if (!regex.test(password)) {
+    res.status(500).render('auth/signup', {
+      errorMessage: 'Password needs to have at least 6 chars and must contain at least one number, one lowercase and one uppercase letter.',
+    });
+    return;
+  }
+
+  try {
+    const salt = await bcryptjs.genSalt(saltRounds);
+    const hashedPassword = await bcryptjs.hash(password, salt);
+    const createdUser = await User.create({ username, passwordHash: hashedPassword });
+    req.login(createdUser, function (err) {
+      if (err) {
+        return next(err);
+      }
+      res.redirect('/dashboard');
+    });
+  } catch (error) {
+    if (error instanceof mongoose.Error.ValidationError) {
+      res.status(500).render('auth/signup', { errorMessage: error.message });
+    } else if (error.code === 11000) {
```

Comment:
what is error code 11000? maybe create a variable that is descriptive instead of the magic number :)